src/main/java/com/refinedmods/refinedstorage/apiimpl/autocrafting/task/v6/calculator/CraftingCalculator.java

```diff
@@ -0,0 +1,362 @@
+
+                    remaining -= toTake;
+
+                    fromSelf = results.get(possibleInput);
+                }
+
+                if (fromNetwork != null && remaining > 0) {
+                    int toTake = Math.min(remaining, fromNetwork.getCount());
+
+                    craftingPreviewInfo.getToTake().add(possibleInput, toTake);
+
+                    node.getRequirements().addItemRequirement(ingredientNumber, possibleInput, toTake, ingredient.getCount());
+
+                    storageSource.remove(fromNetwork, toTake);
+
+                    remaining -= toTake;
+
+                    fromNetwork = storageSource.get(possibleInput);
+
+                    toExtractInitial.add(possibleInput, toTake);
+                }
+
+                if (remaining > 0) {
+                    ICraftingPattern subPattern = network.getCraftingManager().getPattern(possibleInput);
+
+                    if (subPattern != null) {
+                        int qtyPerCraft = getQuantityPerCraft(possibleInput, null, subPattern);
+                        int subQty = ((remaining - 1) / qtyPerCraft) + 1;
+
+                        calculateInternal(subQty, storageSource, fluidStorageSource, results, fluidResults, subPattern, false);
+
+                        fromSelf = results.get(possibleInput);
+                        if (fromSelf == null) {
+                            throw new IllegalStateException("Recursive calculation didn't yield anything");
+                        }
+
+                        fromNetwork = storageSource.get(possibleInput);
+
+                        // fromSelf contains the amount crafted after the loop.
+                        craftingPreviewInfo.getToCraft().add(fromSelf.copy());
+                    } else {
+                        if (!possibleInputs.cycle()) {
+                            // Give up.
+                            possibleInput = possibleInputs.get(); // Revert back to 0.
+
+                            craftingPreviewInfo.getMissing().add(possibleInput, remaining);
+
+                            itemsToExtract.add(possibleInput, remaining);
+
+                            remaining = 0;
+                        } else {
+                            // Retry with new input...
+                            possibleInput = possibleInputs.get();
+
+                            fromSelf = results.get(possibleInput);
+                            fromNetwork = storageSource.get(possibleInput);
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    private void calculateForFluids(int qty,
+                                    IStackList<ItemStack> storageSource,
+                                    IStackList<FluidStack> fluidStorageSource,
+                                    IStackList<ItemStack> results,
+                                    IStackList<FluidStack> fluidResults,
+                                    ICraftingPattern pattern,
```

Comment:
pattern is not used

Reply:
Not true, pattern is used

Reply:
where? I'm not finding it (and intellij isn't either).